apps/src/templates/instructions/ReviewTab.jsx

```diff
@@ -68,56 +70,86 @@ class ReviewTab extends Component {
     // If there's no channelId (happens when a teacher is viewing as a student who has not done any work on a level),
     // do not make API calls that require a channelId
     if (!channelId) {
+      this.setState({
+        initialLoadCompleted: true
+      });
+      this.props.onLoadComplete();
       return;
     }
 
-    codeReviewDataApi
-      .getCodeReviewCommentsForProject(channelId)
-      .done((data, _, request) => {
-        this.setState({
-          comments: data,
-          token: request.getResponseHeader('csrf-token')
-        });
-      });
+    const initialLoadPromises = [];
+
+    initialLoadPromises.push(
+      new Promise((resolve, reject) => {
+        codeReviewDataApi
+          .getCodeReviewCommentsForProject(channelId)
```

Comment:
(likely out of scope for this change)
Is it possible (and would it make sense) to combine these three server calls into one? Since they happen every time the component is mounted, we may be able to save time and server load by making a single call that returns all the needed data.

Reply:
Yeah, good call, I was thinking about this as well - and maybe a slightly bigger question around our paradigms for how we generally generate and assemble data. Do we prefer to keep our controller routes small and specified or do more of the data assembly logic on the back end? Or potentially is there any layer in between where we assimilate server/API data into presentation data? Curious for @bencodeorg's thoughts on this as well.

Reply:
Jessie certainly knows better than I do (especially in * labs), but I feel like we do as little "data assembly" as possible in JS.

That said, I also think we try to put data processing code in models rather than controllers (I've heard the phrase "fat models, skinny controllers" as a common Rails pattern). I think for code review we've tended towards putting some of the data assembly so far in the controllers themselves, which I think is maybe not a typical pattern. In general, I think it's easier to test data assembly stuff when it is in a model vs. controller.